packages/vite/src/node/plugins/resolve.ts

```diff
@@ -487,180 +486,167 @@ function tryFsResolve(
-      options,
-      false,
-      targetWeb,
-      options.tryPrefix,
-      options.skipPackageJson,
-    ))
-  ) {
-    return res
-  }
-
-  if (
-    (res = tryResolveFile(
-      file,
-      postfix,
-      options,
-      false,
-      targetWeb,
-      options.tryPrefix,
-      options.skipPackageJson,
-    ))
-  ) {
-    return res
-  }
-
-  for (const ext of options.extensions) {
-    if (
-      tryUnsplitted &&
-      (res = tryResolveFile(
-        fsPath + ext,
-        '',
-        options,
-        false,
-        targetWeb,
-        options.tryPrefix,
-        options.skipPackageJson,
-        false,
-      ))
-    ) {
-      return res
-    }
-
-    if (
-      (res = tryResolveFile(
-        file + ext,
-        postfix,
+  let postfixIndex = fsPath.indexOf('?')
+  if (postfixIndex < 0) {
+    postfixIndex = fsPath.indexOf('#')
+
+    // Dependencies like es5-ext use `#` in their paths. We don't support `#` in user
+    // source code so we only need to perform the check for dependencies.
+    // We don't support `?` in node_modules paths, so we only need to check in this branch.
+    if (postfixIndex >= 0 && fsPath.includes('node_modules')) {
+      const res = tryCleanFsResolve(
+        fsPath,
         options,
-        false,
+        tryIndex,
         targetWeb,
-        options.tryPrefix,
-        options.skipPackageJson,
-        false,
-      ))
-    ) {
-      return res
+        skipPackageJson,
+      )
+      if (res) return res
     }
   }
 
-  // if `tryIndex` false, skip as we've already tested above
-  if (!tryIndex) return
-
-  if (
-    tryUnsplitted &&
-    (res = tryResolveFile(
-      fsPath,
-      '',
-      options,
-      tryIndex,
-      targetWeb,
-      options.tryPrefix,
-      options.skipPackageJson,
-    ))
-  ) {
-    return res
+  let file = fsPath
+  let postfix = ''
+  if (postfixIndex >= 0) {
+    file = fsPath.slice(0, postfixIndex)
+    postfix = fsPath.slice(postfixIndex)
   }
 
-  if (
-    (res = tryResolveFile(
-      file,
-      postfix,
-      options,
-      tryIndex,
-      targetWeb,
-      options.tryPrefix,
-      options.skipPackageJson,
-    ))
-  ) {
-    return res
-  }
+  const res = tryCleanFsResolve(
+    file,
+    options,
+    tryIndex,
+    targetWeb,
+    skipPackageJson,
+  )
+  if (res) return res + postfix
 }
 
-function tryResolveFile(
+function tryCleanFsResolve(
   file: string,
-  postfix: string,
   options: InternalResolveOptions,
-  tryIndex: boolean,
-  targetWeb: boolean,
-  tryPrefix?: string,
-  skipPackageJson?: boolean,
-  skipTsExtension?: boolean,
+  tryIndex = true,
+  targetWeb = true,
+  skipPackageJson = false,
 ): string | undefined {
-  let stat: fs.Stats | undefined
-  try {
-    stat = fs.statSync(file, { throwIfNoEntry: false })
-  } catch {
-    return
+  const { tryPrefix, extensions, preserveSymlinks } = options
+
+  const fileStat = tryStatSync(file)
+
+  // Try direct match first
+  if (fileStat && !fileStat.isDirectory())
```

Comment:
I know we also use `!stat.isDirectory()` before but maybe now we can use 
```suggestion
  if (fileStat?.isFile())
```
so it's more direct

Reply:
Nice, looks a lot better 👍🏼